pandas/io/formats/csvs.py

```diff
@@ -127,14 +128,19 @@ def save(self):
         else:
             encoding = self.encoding
 
-        if hasattr(self.path_or_buf, 'write'):
-            f = self.path_or_buf
-            close = False
+        # PR 21300 uses string buffer to receive csv writing and dump into
+        # file-like output with compression as option. GH 21241, 21118
+        f = StringIO()
+        if not is_file_like(self.path_or_buf):
+            # path_or_buf is path
+            path_or_buf = self.path_or_buf
+        elif hasattr(self.path_or_buf, 'name'):
```

Comment:
Is this actually required? If we already have a handle aren't we just round tripping here anyway by converting the handle to a name then back to a handle?

Reply:
Hi, I think so it's required because we need to generate a new handle from _get_handle potentially with compression and encoding requirement) when a external handle is passed. currently when a file handle is passed, no compression is done even when specified https://github.com/pandas-dev/pandas/issues/21227

Reply:
Aren't we leaving a file handle open then? To illustrate:

```python
>>> f1 = open('foo.txt')
>>> name = f1.name
>>> f2, handles = _get_handle(name, 'r')
>>> f1.closed
False
>>> f2.closed
False
```

AFAICT with your code you eventually would close the second object but the first would stay open, no?

Reply:
Hmm OK I see your point, but do we have test coverage for all of the branches? It just seems a little strange to me that we are opening two handles to the same location when a handle gets passed to this function. I feel like that's asking for some strange behavior to pop up, but would feel more comfortable if the test coverage was there to hit all three branches. lmk

Reply:
see your point. though two file handles have separate BufferIO and states. added tests using file-handle as input and assert equivalence. we also have compress_size test with file-handle showing they fix #21227 

we already have tests for path and buffer IO as they are existing use cases.

Reply:
According to the issue #21561 we have a bug if the file is `sys.stdout`. Basically the library creates a file named `<stdout>`and write stuff in it